modular/gater/object_handler.go

```diff
@@ -198,33 +199,45 @@ func (g *GateModular) getObjectHandler(w http.ResponseWriter, r *http.Request) {
 		err = ErrInvalidRange
 		return
 	}
-	var low int64
-	var high int64
+
 	if isRange {
-		low = rangeStart
-		high = rangeEnd
+		lowOffset = rangeStart
+		highOffset = rangeEnd
 	} else {
-		low = 0
-		high = int64(objectInfo.GetPayloadSize()) - 1
+		lowOffset = 0
+		highOffset = int64(objectInfo.GetPayloadSize()) - 1
 	}
 
 	task := &gfsptask.GfSpDownloadObjectTask{}
 	task.InitDownloadObjectTask(objectInfo, bucketInfo, params, g.baseApp.TaskPriority(task), reqCtx.Account(),
-		low, high, g.baseApp.TaskTimeout(task, uint64(high-low+1)), g.baseApp.TaskMaxRetry(task))
-	data, err := g.baseApp.GfSpClient().GetObject(reqCtx.Context(), task)
-	if err != nil {
+		lowOffset, highOffset, g.baseApp.TaskTimeout(task, uint64(highOffset-lowOffset+1)), g.baseApp.TaskMaxRetry(task))
+	if pieceInfos, err = downloader.SplitToSegmentPieceInfos(task, g.baseApp.PieceOp()); err != nil {
 		log.CtxErrorw(reqCtx.Context(), "failed to download object", "error", err)
 		return
 	}
 	w.Header().Set(ContentTypeHeader, objectInfo.GetContentType())
 	if isRange {
-		w.Header().Set(ContentRangeHeader, "bytes "+util.Uint64ToString(uint64(low))+
-			"-"+util.Uint64ToString(uint64(high)))
+		w.Header().Set(ContentRangeHeader, "bytes "+util.Uint64ToString(uint64(lowOffset))+
+			"-"+util.Uint64ToString(uint64(highOffset)))
 	} else {
 		w.Header().Set(ContentLengthHeader, util.Uint64ToString(objectInfo.GetPayloadSize()))
 	}
-	w.Write(data)
-	log.CtxDebugw(reqCtx.Context(), "succeed to download object")
+	for idx, pInfo := range pieceInfos {
+		enableCheck := false
+		if idx == 0 { // only check in first piece
+			enableCheck = true
+		}
+		pieceTask := &gfsptask.GfSpDownloadPieceTask{}
+		pieceTask.InitDownloadPieceTask(objectInfo, bucketInfo, params, g.baseApp.TaskPriority(task),
+			enableCheck, reqCtx.Account(), uint64(highOffset-lowOffset+1), pInfo.SegmentPieceKey, pInfo.Offset,
```

Comment:
highOffset and lowOffset comes from object payload data size.
but get piece size should come from pieceInfo?

Reply:
Yes, `hightOffset/lowOffset` is the object‘s property used to check quota. And `pInfo.Offset/pInfo.Length` is the piece's property used to reserve memory resources.